### DISCUSSION TITLE
(React) Question about updating options real time without re-rendering the component

### DISCUSSION THREAD

**Discussion by SilasCundiff** [matteobruni/tsparticles#1148] [Q&A] [answered]
My apologies if this is the wrong place to post about this, or if this has already been answered and I did not find it. If it is the wrong place, please redirect me to the correct location!
**TLDR at bottom.**

This is my first non-tutorial based react project, so this question is most likely just due to a lack of understanding of core concepts on my part. I am very hesitant to ask questions that might have been answered elsewhere, but I have searched for an explanation for a few days now and have not found anything that helped me understand how to fix the problem.

#### Background of my project:
 I am attempting to use the React particles library to enhance an audio visualizer. I essentially want the speed of the particles to update real time with the tempo of the music, which I have achieved to some capacity.

#### The issue as I, maybe mistakenly, understand it: 
I am storing the particle options I need to update in state, and they are updating properly with frequency array I am using to calculate the speed of the particles. However, unless I am mistaken, this method of storing the options in state and updating them every tick is causing the component to re-render, causing massive performance issues and crashes.

I have read that you can use useRef to update a canvas without re-rendering a canvas, but I am not experienced enough to know how to do that with a component that someone else wrote.

#### TLDR: Essentially, my question is: 

1. Can you update the react-tsparticle component’s options without invoking a re-render and have the changes take effect on the canvas in real time?
2. If what I am attempting is feasible, what should I research to learn a solution and improve my understanding?
3. Am I just completely wrong about my assumptions and asking the wrong questions?

I want to improve my understanding so that I don't encounter the same road block in the future. so Any information or direction would be greatly appreciated! This issue is the one thing that has prevented me from moving on with the project for the past several days.

**Comment by matteobruni** [comment, marked as answer]
This can be tricky to achieve since the reload is necessary for adjusting the retina sizes and distances (for retina displays).

But you can write a `noise` plugin to override the particle speed.

You can see a sample Perlin noise script here: https://github.com/matteobruni/tsparticles/blob/master/demo/main/public/javascripts/noise.js

And its usage here: https://github.com/matteobruni/tsparticles/blob/4340750b22588bc764845f58b2848377b20ca9ee/demo/main/public/javascripts/demo.js#L116

To retrieve the particles container object using the react library you can read more here: https://github.com/matteobruni/tsparticles/issues/1108#issuecomment-740035238

Since this can be difficult if you have any more questions don't hesitate to ask, it's a great idea and I'd like to share it!

**Reply by SilasCundiff** [reply]
Okay, thank you! I will look into the resources you posted, and will post back if I get to stuck. If I can figure out how to implement this, I will definately post a link to the project once it's complete in case others wanted to do something similar.

**Reply by SilasCundiff** [reply]
You weren’t kidding when you said that this is tricky. Thank you for suggesting Perlin Noise because it has introduced me to some awesome concepts that I didn’t know I wanted to learn. Perlin Noise was something I didn’t know existed until yesterday, so it’s a lot to digest.

I wanted to ask a few questions just to clarify some things, to make sure I’m on the right track. I looked at the code examples you provided on noise, and it’s usage, and I think I understand the most basic idea of it. I also found your Codepens that utilize a noise field to influence the direction of the particles.

So, if I’m not mistaken, the idea is to use a noise field to manipulate the speed of the particles without directly altering the particle’s options, perhaps by creating obstacles to slow them down?

If that is the case, then I would update NoiseZ using the audio data to influence the field in the update method, and then in the generate method is where the particles are influenced using the fields data?

This project is a lot more complex than I thought it would be, and I’ve never written a plugin before, although I think I can figure it out with the help of Google. Maybe I should have started with a TODO app as my first project instead, but I really love the tsparticle library and I’m invested in at least attempting to make this idea work.

**Reply by matteobruni** [reply]
Well perlin and simplex noise are cool effects to implement, but what I wanted to show you is the `setNoise` function to alter the particle speed in an easy way. You just need to return an object like the one used in the sample I provided.

You have full particle access and you can alter all the speed options, you can also set `horizontal` and `vertical` velocities to `0` and return your preferred values so they're applied. Remember that those values are clamped between `-1` and `1`.

I made a simpler sample here: https://codepen.io/matteobruni/pen/rNMyPPL

Hope it helps more, without distracting with other cool algorithms 😅

**Reply by SilasCundiff** [reply]
Over analyzing is my signature move. Regardless, I have learned a lot from it and now have some nightly reading material.

Thank you for taking the time to set up a simplified version, I’ve been playing around with it in codepen since you posted it and I think I now have a better understanding of how to update the options through it.

I was able to update the speed and a few other options in codepen, now I just need to take the time to try to implement something similar in my own project.

**Reply by SilasCundiff** [reply]
I got it working! So, I’ll mark this as answered. And it was as simple as you said, it was just my lack of understanding that really caused me issues. Thank you so much! You have saved me a ton of time and frustration, and I’ve learned so much about how object references work in react after tinkering around with it. It’s true tutorials can’t touch projects when it comes to comprehending concepts.

I won’t post a link just yet, as the project is far from finished. For instance, the particles change speed too fast currently, because of how quickly my frequencyArray updates. I need to work out how to smoothen those updates out, but that’s unrelated to tsparticles, so I won’t bother you with it. 

Once I’m done with the project, I’ll post it in the show & tell section. Possibly as a codepen, so others can play with it, if I can figure out how codepen works.

To hopefully pay it forward a bit, 
Instructions for anyone else that wants to do this in react and is a noob like me:

Make sure that your particles have `noise {enabled: true}` in the initial particle options, otherwise the particles seem to ignore your setNoise call. This took me an embarrassing amount of time to figure out.
1.	Create a reference to the particles container using the provided `container` prop
2.	create a function that takes the container as an argument – I named mine `setContainer(container)`
3.	call it in your componantDidMount, or useEffect (I believe the container needs to refresh after setNoise is called, so don't call this too often or it will result in a lot of flashing from all the refreshes)
4.	call the `container.setNoise({...})` method inside of setContainer
5.	setNoise contains an object with init, update, and generate methods. Generate is the one I used to alter particle options.
6.	Pass your particle as an argument to the generate method - `generate: function (particle){...}`
7.	From inside of generate you can access `particle.updater.particle.movespeed`, or any of the other options, and alter them with your new values.

Reference the codepen matteobruni posted for a more completed look.

**Reply by matteobruni** [reply]
Instead of `particle.updater.particle.moveSpeed` you can just update `particle.moveSpeed` since the `updater.particle` is the `particle` itself.

I'm really curious about the result, I hope you'll show us in the show and tell.

Anyway the `noise` options have a `delay` property too, you can see it here: https://particles.js.org/interfaces/_options_interfaces_particles_move_noise_inoise_.inoise.html
